Database Failures in Subpanel (MySQL Error)

<!--- Provide a general summary of the issue in the **Title** above -->
<!--- Before you open an issue, please check if a similar issue already exists or has been closed before. --->
<!--- If you have discovered a security risk please report it by emailing security@suitecrm.com. This will be delivered to the product team who handle security issues. Please don't disclose security bugs publicly until they have been handled by the security team. --->
<!--- Please be aware that as of the 31st January 2022 we no longer support 7.10.x.
New issues referring to 7.10.x will only be valid if applicable to 7.12.x and above.
If your issue is still applicable in 7.12.x, please create the issue following the template below -->
#### Issue
<!--- Provide a more detailed introduction to the issue itself, and why you consider it to be a bug -->
Database error in subpanels with the error:  ERROR 1250 (42000): Table ‘…’ from one of the SELECTs cannot be used in global ORDER clause.  The issue pertains to use the Table name in the sort clause.  This seemed to work fine until today (Oct. 10. 2022).  So must have been a recent update in mySQL that no longer permits this syntax.  I was able to get the subpanel to display contacts again by sorting on a different column.  Example below (screenshots) is in the Cases module and Contacts Subpanel.  However, this issue is throughout  SuiteCRM and not just this module, and not just contacts.
![sort problem](https://user-images.githubusercontent.com/29544935/195383235-d940e040-4b12-495b-8b7f-bb7f0afa4496.png)
When I sort on "Accounts" it works:
![working sort problem](https://user-images.githubusercontent.com/29544935/195383318-aff4c101-e63a-44e1-8fd7-2713a6eb00e2.png)

<!--- Ensure that all code ``` is surrounded ``` by triple back quotes. This can also be done over multiple lines -->

#### Expected Behavior
<!--- Tell us what should happen -->
The contact in the subpanel should display without error.

#### Actual Behavior
<!--- Tell us what happens instead -->
Contacts in subpanel display with database error.
<!--- Also please check relevant logs (suitecrm.log, php error.log etc.) -->```Wed Oct 12 14:08:16 2022 [16187][1][FATAL] Error retrieving Case list:  Query Failed: (SELECT contacts.id ,  contacts.first_name ,  contacts.last_name ,  contacts.salutation ,  LTRIM(RTRIM(CONCAT(IFNULL(contacts.first_name,''),' ',IFNULL(contacts.last_name,'')))) as name,  '                                                                                                                                                                                                                                                              ' account_name ,  '                                    '  account_id ,  contacts.phone_work ,  contacts.assigned_user_id  ,  'contacts' panel_name  FROM contacts  LEFT JOIN contacts_cstm ON contacts.id = contacts_cstm.id_c  INNER JOIN  contacts_cases ON contacts.id=contacts_cases.contact_id AND contacts_cases.case_id='9e5568fd-3fd7-5130-0a83-6345f29aafeb' AND contacts_cases.deleted=0

  where contacts.deleted=0) ORDER BY contacts.last_name, contacts.first_name asc LIMIT 0,10: MySQL error 1250: Table 'contacts' from one of the SELECTs cannot be used in global ORDER clause ``` 

#### Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->
As a work around, I can sort by another column to avoid the error.

The fix is to remove  the table name from "sort by" I think, but this is throughout all the subpanels not sure where to start.  the ORDERBY contacts.last_name I think is what's causing the problem and should be just ORDERBY last_name.

#### Steps to Reproduce
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug include code to reproduce, if relevant -->
1.  Go into any module that has Contacts subpanel, see it doesn't work.  Also Project Tasks have same issue so its more than just contacts.  See screen shots provided.


#### Context
<!--- How has this bug affected you? What were you trying to accomplish? -->
<!--- If you feel this should be a low/medium/high priority then please state so -->
This is a huge issue, I imagine affects every installation on MySQL 8.0.31

#### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* SuiteCRM Version used:  7.12.7
* Browser name and version (e.g. Chrome Version 51.0.2704.63 (64-bit)):  Chrome
* Environment name and version (e.g. MySQL, PHP 7): MySQL 8.0.31
* Operating System and version (e.g Ubuntu 16.04):  WHM/Cpanel, Centoi v7.9.2009

I set level to ERROR and Enable stack trace to checked and it doesn't give me any further info.  This is the output.....

Thu Oct 20 14:12:58 2022 [20457][1][FATAL] Error retrieving Case list:  Query Failed: (SELECT contacts.id ,  contacts.first_name ,  contacts.last_name ,  contacts.salutation ,  LTRIM(RTRIM(CONCAT(IFNULL(contacts.first_name,''),' ',IFNULL(contacts.last_name,'')))) as name,  '                                                                                                                                                                                                                                                              ' account_name ,  '                                    '  account_id ,  contacts.phone_work ,  contacts.assigned_user_id  ,  'contacts' panel_name  FROM contacts  LEFT JOIN contacts_cstm ON contacts.id = contacts_cstm.id_c  INNER JOIN  contacts_cases ON contacts.id=contacts_cases.contact_id AND contacts_cases.case_id='6a66b4d6-34a3-ef97-0de0-634878effa1e' AND contacts_cases.deleted=0

  where contacts.deleted=0) ORDER BY contacts.last_name, contacts.first_name asc LIMIT 0,10: MySQL error 1250: Table 'contacts' from one of the SELECTs cannot be used in global ORDER clause
Thu Oct 20 14:12:58 2022 [20457][1][FATAL] [ListView::processUnionBeans] . Database failure. Please refer to suitecrm.log for details.
Thu Oct 20 14:12:59 2022 [20457][1][ERROR] Unable to find relationship emails_email_templates
Thu Oct 20 14:12:59 2022 [20457][1][ERROR] Unable to find relationship emails_email_templates
Thu Oct 20 14:13:54 2022 [20629][1][FATAL] Mysqli_query failed.
Thu Oct 20 14:13:54 2022 [20629][1][FATAL] Error retrieving Case list:  Query Failed: (SELECT contacts.id ,  contacts.first_name ,  contacts.last_name ,  contacts.salutation ,  LTRIM(RTRIM(CONCAT(IFNULL(contacts.first_name,''),' ',IFNULL(contacts.last_name,'')))) as name,  '                                                                                                                                                                                                                                                              ' account_name ,  '                                    '  account_id ,  contacts.phone_work ,  contacts.assigned_user_id  ,  'contacts' panel_name  FROM contacts  LEFT JOIN contacts_cstm ON contacts.id = contacts_cstm.id_c  INNER JOIN  contacts_cases ON contacts.id=contacts_cases.contact_id AND contacts_cases.case_id='6a66b4d6-34a3-ef97-0de0-634878effa1e' AND contacts_cases.deleted=0

  where contacts.deleted=0) ORDER BY contacts.last_name, contacts.first_name asc LIMIT 0,10: MySQL error 1250: Table 'contacts' from one of the SELECTs cannot be used in global ORDER clause
Thu Oct 20 14:13:54 2022 [20629][1][FATAL] [ListView::processUnionBeans] . Database failure. Please refer to suitecrm.log for details.


There's nothing in the suitecrm.log

--- a/data/SugarBean.php
+++ b/data/SugarBean.php
@@ -841,7 +841,8 @@ class SugarBean
         }
 
 
-        $first = true;
+        $final_query_array = [];
+        $final_query_rows_array = [];
 
         //Breaking the building process into two loops. The first loop gets a list of all the sub-queries.
         //The second loop merges the queries and forces them to select the same number of columns
@@ -869,17 +870,23 @@ class SugarBean
                 } else {
                     $tmp_final_query = $parentbean->$shortcut_function_name();
                 }
-                if (!$first) {
-                    $final_query_rows .= ' UNION ALL ( '
-                        . $parentbean->create_list_count_query($tmp_final_query, $parameters) . ' )';
-                    $final_query .= ' UNION ALL ( ' . $tmp_final_query . ' )';
-                } else {
-                    $final_query_rows = '(' . $parentbean->create_list_count_query($tmp_final_query, $parameters) . ')';
-                    $final_query = '(' . $tmp_final_query . ')';
-                    $first = false;
+
+                if (!empty($tmp_final_query)) {
+                    $final_query_array[] = $tmp_final_query;
+                    $final_query_rows_array[] = $parentbean->create_list_count_query($tmp_final_query, $parameters);
                 }
             }
         }
+
+        if (count($final_query_array) > 1) {
+            $final_query =  '( ' . implode(' ) UNION ALL ( ', $final_query_array) . ' )';
+            $final_query_rows =  '( ' . implode(' ) UNION ALL ( ', $final_query_rows_array) . ' )';
+        } elseif (count($final_query_array) === 1) {
+            $final_query = $final_query_array[0];
+            $final_query_rows = $final_query_rows_array[0];
+        }
+        $queryFromDatasourceFunction = !empty($final_query);
+
         //If final_query is still empty, its time to build the sub-queries
         if (empty($final_query)) {
             $subqueries = SugarBean::build_sub_queries_for_union($subpanel_list, $subpanel_def, $parentbean, $order_by);
@@ -893,7 +900,8 @@ class SugarBean
                 }
                 $subqueries[$i]['query_fields'] = $query_fields;
             }
-            $first = true;
+            $final_query_array = [];
+            $final_query_rows_array = [];
             //Now ensure the queries have the same set of fields in the same order.
             foreach ($subqueries as $subquery) {
                 $subquery['select'] = "SELECT";
@@ -925,13 +933,6 @@ class SugarBean
 
                 //Put the query into the final_query
                 $query = $subquery['select'] . " " . $subquery['from'] . " " . $subquery['where'];
-                if (!$first) {
-                    $query = ' UNION ALL ( ' . $query . ' )';
-                    $final_query_rows .= " UNION ALL ";
-                } else {
-                    $query = '(' . $query . ')';
-                    $first = false;
-                }
                 $query_array = $subquery['query_array'];
                 $select_position = strpos((string) $query_array['select'], "SELECT");
                 $distinct_position = strpos((string) $query_array['select'], "DISTINCT");
@@ -959,8 +960,17 @@ class SugarBean
                     }
                     $secondary_queries[] = $subquerystring;
                 }
-                $final_query .= $query;
-                $final_query_rows .= $query_rows;
+
+                $final_query_array[] = $query;
+                $final_query_rows_array[] = $query_rows;
+            }
+
+            if (count($final_query_array) > 1) {
+                $final_query =  '( ' . implode(' ) UNION ALL ( ', $final_query_array) . ' )';
+                $final_query_rows =  '( ' . implode(' ) UNION ALL ( ', $final_query_rows_array) . ' )';
+            } elseif (count($final_query_array) === 1) {
+                $final_query = $final_query_array[0];
+                $final_query_rows = $final_query_rows_array[0];
             }
         }
 
@@ -982,6 +992,17 @@ class SugarBean
 
             $order_by = $parentbean->process_order_by($order_by, $submodule, $suppress_table_name);
             if (!empty($order_by)) {
+                if ($queryFromDatasourceFunction) {
+                    // Strip any trailing ORDER BY (and any LIMIT that follows it) embedded by the datasource function
+                    // before appending the requested sort. It only matches the outermost clause, ORDER BY inside a
+                    // parenthesised subquery is always followed by an unmatched ")" and won't be touched. Supports
+                    // function calls with up to 2 levels of nested parens (e.g. FIELD(), COALESCE()).
+                    $final_query = preg_replace(
+                        '/\s+ORDER\s+BY\b(?:[^()]|\((?:[^()]*|\([^()]*\))*\))*$/i',
+                        '',
+                        $final_query
+                    );
+                }
                 $final_query .= ' ORDER BY ' . $order_by;
             }
         }

